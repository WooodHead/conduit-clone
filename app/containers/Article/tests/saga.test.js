/**
 * Test  sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { takeLatest, put } from 'redux-saga/effects';
import {
  FETCH_ARTICLE,
  GET_ARTICLE_SLUG,
} from 'containers/Article/constants';

import {
  articleLoadingError,
  commentsLoadingError,
} from 'containers/Article/actions';

import articleData, { getArticle, getComments } from '../saga';

const payload = {
  article: {
    foo: 'bar',
  },
  comments: [{
    foo: 'bar',
  }],
};
const slug = 'sample-slug';

describe('getArticle', () => {
  let getArticleGenerator;

  beforeEach(() => {
    getArticleGenerator = getArticle({
      slug,
    });

    const callDescriptor = getArticleGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();

    let putDescriptor = getArticleGenerator.next(slug).value;
    expect(putDescriptor).toMatchSnapshot();

    putDescriptor = getArticleGenerator.next(payload.article).value;
    expect(putDescriptor).toMatchSnapshot();
  });

  it('should dispatch the articleLoadingError action if api requests failed', () => {
    const putDescriptor = getArticleGenerator.throw().value;
    expect(putDescriptor).toEqual(put(articleLoadingError()));
  });
});

describe('getComments', () => {
  let getCommentsGenerator;

  beforeEach(() => {
    getCommentsGenerator = getComments({
      slug,
    });

    const selectDescriptor = getCommentsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getCommentsGenerator.next(payload.comments).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the commentsLoadingError action if api requests failed', () => {
    const putDescriptor = getCommentsGenerator.throw().value;
    expect(putDescriptor).toEqual(put(commentsLoadingError()));
  });
});

describe('articlesDataSaga Saga', () => {
  const articleDataSaga = articleData();

  it('should start task to watch for FETCH_ARTICLES action with getArticle', () => {
    const takeLatestDescriptor = articleDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(FETCH_ARTICLE, getArticle));
  });

  it('should start task to watch for GET_ARTICLE_SLUG action with getComments', () => {
    const takeLatestDescriptor = articleDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(GET_ARTICLE_SLUG, getComments));
  });
});
