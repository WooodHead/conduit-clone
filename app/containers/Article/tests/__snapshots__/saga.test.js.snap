// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getArticle should dispatch the articleLoadingError action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://conduit.productionready.io/api/articles/sample-slug",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getArticle should dispatch the articleLoadingError action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "article": undefined,
      "type": "app/Article/ARTICLE_LOADED",
    },
    "channel": null,
  },
}
`;
